src/hardware/mixer.cpp

```diff
@@ -2351,49 +2340,61 @@ MixerChannel::~MixerChannel()
 	}
 }
 
-static constexpr float calc_frames_per_tick(const int freq_hz)
+// We use floats in the range of 16 bit integers everywhere.
+// SDL expects floats to be normalized from 1.0 to -1.0
+// It might be better for us to use normalized floats elsewhere in the future.
+// For now, that probably breaks some assumptions elsewhere in the mixer.
+// So just normalize as a final step before sending the data to SDL.
+static float normalize_sample(float sample)
 {
-	assert(freq_hz > 0);
-
-	constexpr auto MillisecondsPerTick = 1000;
-	return static_cast<float>(freq_hz) / MillisecondsPerTick;
+	return sample / 32768.0f;
 }
 
 // Mix a certain amount of new sample frames
-static void mix_samples(const int frames_requested)
+std::vector<AudioFrame> mix_samples(const int frames_requested)
 {
 	assert(frames_requested >= 0);
 
-	constexpr auto CaptureBufFrames = 1024;
-
-	const auto frames_added = std::min(frames_requested - mixer.frames_done,
-	                                   CaptureBufFrames);
-
-	const auto pos_offset     = mixer.pos + mixer.frames_done;
-	const auto start_work_pos = mixer.work.begin() + pos_offset;
+	std::vector<AudioFrame> output_buffer(frames_requested);
+	std::vector<AudioFrame> reverb_buffer(frames_requested);
+	std::vector<AudioFrame> chorus_buffer(frames_requested);
```

Comment:
These temporary vectors are what @interloper98 was seeing in those measurements.  These would be easy to refactor to stop allocating every block.  The SoundBlaster `AudioVector` is more tricky...

Reply:
These are removed now, although it turned out other temporary vectors in `AddSamples` and `ConvertSamplesAndMaybeZohUpsample` were allocating more than these.  Those are removed now as well.

Reply:
Noice 😎